### PR TITLE
feat: TLS listeners supported

### LINKED ISSUE #64: Enable TLS listeners on `Gateway`s
## Problem Statement

KGO 1.2 allows HTTP and HTTPS listeners only. KIC already supports `TLSRoute`s, therefore we need to enable users to create `Gateway`s with `TLS` listeners with `Passthrough` TLS mode enabled.

## Proposed Solution

- In case a Gateway with TLS listener is created, it is accepted (only in case the TLS mode is `Passthrough`)
- Enable the `GatewayAlpha` FG in the `ControlPlane`
- The `DataPlane` gets the STREAM_LISTEN env var set accordingly
- The `DataPlane` Ingress service gets configured to expose the TLS port of the `DataPlane` 

## Acceptance Criteria
- [ ] As a user, I can create a `Gateway` with a TLS listener and use `TLSRoute`s to configure the `DataPlane`

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:

Support for TLS Listeners and TLSRoutes.

**Which issue this PR fixes**

Fixes #64 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: controller/gateway/controller_conditions.go [hunk 1]
```diff
@@ -32,4 +32,10 @@ const (
 	// to express that more than one TLS secret has been set in the listener
 	// TLS configuration.
 	ListenerReasonTooManyTLSSecrets k8sutils.ConditionReason = "TooManyTLSSecrets"
+
+	// ListenereReasonInvalidTLSMode must be used with the Accepted condition
+	// to express that the listener has an invalid TLS mode.
+	// HTTPS can only be configured with mode Terminate, while TLS can only be
+	// be configured with mode Passthrough.
+	ListenereReasonInvalidTLSMode k8sutils.ConditionReason = "InvalidTLSMode"
```
Comment (pmalek) [inline controller/gateway/controller_conditions.go:40]:
```suggestion
	// ListenerReasonInvalidTLSMode must be used with the Accepted condition
	// to express that the listener has an invalid TLS mode.
	// HTTPS can only be configured with mode Terminate, while TLS can only be
	// be configured with mode Passthrough.
	ListenerReasonInvalidTLSMode k8sutils.ConditionReason = "InvalidTLSMode"
```

We might add a reference link to https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls which contains the table with allowed modes for particular route types.

File: controller/controlplane/controller_utils_test.go [hunk 2]
```diff
@@ -434,6 +434,14 @@ func TestSetControlPlaneDefaults(t *testing.T) {
 											Name:  "CONTROLLER_ADMISSION_WEBHOOK_LISTEN",
 											Value: consts.ControlPlaneAdmissionWebhookEnvVarValue,
 										},
+										{
+											Name:  "",
+											Value: consts.ControlPlaneAdmissionWebhookEnvVarValue,
+										},
```
Comment (pmalek) [inline controller/controlplane/controller_utils_test.go:440]: I believe we don't need this right?

File: controller/gateway/controller_reconciler_utils.go [hunk 3]
```diff
@@ -590,10 +590,25 @@ func (g *gatewayConditionsAndListenersAwareT) setAccepted() {
 			LastTransitionTime: metav1.Now(),
 			ObservedGeneration: g.Generation,
 		}
-		if listener.Protocol != gatewayv1.HTTPProtocolType && listener.Protocol != gatewayv1.HTTPSProtocolType {
+		if listener.Protocol != gatewayv1.HTTPProtocolType &&
+			listener.Protocol != gatewayv1.HTTPSProtocolType &&
+			listener.Protocol != gatewayv1.TLSProtocolType {
```
Comment (pmalek) [inline controller/gateway/controller_reconciler_utils.go:595]: Can we use the `supportedRoutesByProtocol()` like so:

```
_, ok := supportedRoutesByProtocol()[listener.Protocol]
```

and then check `!ok`? This way we can avoid forgetting to update this when we add support for other protocols. WDYT?

File: controller/gateway/controller_reconciler_utils.go [hunk 4]
```diff
@@ -590,10 +590,25 @@ func (g *gatewayConditionsAndListenersAwareT) setAccepted() {
...
+			acceptedCondition.Reason = string(ListenereReasonInvalidTLSMode)
+			acceptedCondition.Message = "Only Terminate mode is supported with HTTPS listeners"
+		}
+
+		// Only TLS passthrough mode supported with TLS	Listeners.
+		if listener.Protocol == gatewayv1.TLSProtocolType && *listener.TLS.Mode != gatewayv1.TLSModePassthrough {
+			acceptedCondition.Status = metav1.ConditionFalse
+			acceptedCondition.Reason = string(ListenereReasonInvalidTLSMode)
+			acceptedCondition.Message = "Only Passthrough mode is supported with TLS listeners"
+		}
```
Comment (pmalek) [inline controller/gateway/controller_reconciler_utils.go:611]: WDYT about using a `switch` on the protocol here and then nesting the particular conditions for those protocols inside the `case`s?

File: config/samples/gateway-with-multiple-listeners.yaml
Comment (pmalek) [inline config/samples/gateway-with-multiple-listeners.yaml:1]: Can we add one more sample with just the `TLSRoute` and a `Gateway` with `TLS` listener?

**Comment by mlavacca** [comment]
Closing as not planned for now

**Pull request closed without merging**